ipa-step/src/gate.rs

```diff
@@ -54,50 +54,57 @@ pub fn build<S: CompactStep>() {
     };
     let name_maker = GateName::new(name);
     let gate_name = name_maker.name();
-
     let out = PathBuf::from(env::var("OUT_DIR").unwrap()).join(name_maker.filename());
     println!("writing Gate implementation {gate_name} (for {step_name}) to {out:?}");
+    let gate_impl = compact_gate_impl::<S>(&gate_name);
+
+    write(out, prettyplease::unparse(&parse2(gate_impl).unwrap())).unwrap();
+}
+
+fn compact_gate_impl<S: CompactStep>(gate_name: &str) -> TokenStream {
+    let ident: Ident = parse_str(gate_name).unwrap();
 
     let mut step_narrows = HashMap::new();
     step_narrows.insert(std::any::type_name::<S>(), vec![0]); // Add the first step.
-    let mut as_ref_arms = TokenStream::new();
-    let mut from_arms = TokenStream::new();
 
-    let ident: Ident = parse_str(&gate_name).unwrap();
+    let step_count = usize::try_from(S::STEP_COUNT).unwrap();
+    // this is an array of gate names indexed by the compact gate index.
+    let mut gate_names = Vec::with_capacity(step_count);
+    // builds a lookup table for steps to resolve them to the compact gate index.
+    let mut step_hasher = HashingSteps::new(&ident);
+
     for i in 1..=S::STEP_COUNT {
         let s = String::from("/") + &S::step_string(i - 1);
-        as_ref_arms.extend(quote! {
-            #i => #s,
-        });
-        from_arms.extend(quote! {
-            #s => Ok(#ident(#i)),
-        });
+        step_hasher.hash(&s, i);
+        gate_names.push(s);
+
         if let Some(t) = S::step_narrow_type(i - 1) {
             step_narrows.entry(t).or_insert_with(Vec::new).push(i);
         }
     }
 
     let from_panic = format!("unknown string for {gate_name}: \"{{s}}\"");
+    let gate_lookup_type = step_hasher.lookup_type();
     let mut syntax = quote! {
+
+        static STR_LOOKUP: [&str; #step_count] = [#(#gate_names),*];
+        static GATE_LOOKUP: #gate_lookup_type = #step_hasher
+
         impl ::std::convert::AsRef<str> for #ident {
-            #[allow(clippy::too_many_lines)]
             fn as_ref(&self) -> &str {
-                match self.0 {
+                match usize::try_from(self.0).unwrap() {
                     0 => "/",
-                    #as_ref_arms
-                    _ => unreachable!(),
+                    i => STR_LOOKUP[i - 1],
```

Comment:
I did that and reverted that change eventually because the way it is written now shows what is actually stored inside `STR_LOOKUP` and `GATE_LOOKUP`. I don't strongly object it, just feels that it is more explanatory